src/passes/inlineAssemblyTransformer.ts

```diff
@@ -0,0 +1,176 @@
+
+  verifyNode(node: YulNode) {
+    const transformerName = generateTransformerName(node.nodeType);
+    if (!(transformerName in YulTransformer.prototype))
+      this.errors.push(`${node.nodeType} is not supported`);
+
+    const verifierName = `check${node.nodeType}`;
+    const method = this[verifierName as keyof YulVerifier] as (node: YulNode) => void;
+    if (method !== undefined) return method.bind(this)(node);
+  }
+  checkYulAssignment(node: YulNode) {
+    this.verifyNode(node.value);
+  }
+
+  checkYulFunctionCall(node: YulNode) {
+    if (binaryOps[node.functionName.name] === undefined)
+      this.errors.push(`${node.functionName.name} is not supported`);
+    node.arguments.forEach((arg: YulNode) => this.verifyNode(arg));
+  }
+}
+
+class YulTransformer {
+  vars: Map<string, VariableDeclaration>;
+  ast: AST;
+
+  constructor(ast: AST, externalReferences: any[], context: ASTContext) {
+    this.ast = ast;
+    const blockRefIds = externalReferences.map((ref) => ref.declaration);
+    const blockRefs = [...context.map].filter(([id, _]) => blockRefIds.includes(id)) as [
+      number,
+      VariableDeclaration,
+    ][];
+    this.vars = new Map(blockRefs.map(([_, ref]) => [ref.name, ref]));
+  }
+
+  toSolidityStatement(node: YulNode): Statement {
+    return this.toSolidityNode(node);
+  }
+
+  toSolidityExpr(node: YulNode, ...args: unknown[]): Expression {
+    return this.toSolidityNode(node, ...args) as Expression;
+  }
+
+  toSolidityNode(node: YulNode, ...args: unknown[]): ASTNode {
+    const methodName = generateTransformerName(node.nodeType);
+    const method = this[methodName as keyof YulTransformer] as (node: YulNode) => ASTNode;
+    return method.bind(this)(node, ...(args as []));
+  }
+
+  transformYulLiteral(node: YulNode): Literal {
+    return createNumberLiteral(node.value, this.ast);
+  }
+
+  transformYulIdentifier(node: YulNode): Identifier {
+    const variableDeclaration = this.vars.get(node.name);
+    assert(variableDeclaration !== undefined, `Variable ${node.name} not found.`);
+    return createIdentifier(variableDeclaration, this.ast);
+  }
+
+  transformYulFunctionCall(node: YulNode, typeString: string): BinaryOperation {
+    const leftExpr = this.toSolidityExpr(node.arguments[0], typeString);
+    const rightExpr = this.toSolidityExpr(node.arguments[1], typeString);
+    return new BinaryOperation(
+      this.ast.reserveId(),
+      node.src,
+      typeString,
+      binaryOps[node.functionName.name],
+      leftExpr,
+      rightExpr,
+      node,
+    );
+  }
+
+  transformYulAssignment(node: YulNode): ExpressionStatement {
+    let lhs: Expression;
+    if (node.variableNames.length === 1) {
+      lhs = this.transformYulIdentifier(node.variableNames[0]);
+    } else {
+      lhs = createTuple(
+        this.ast,
+        node.variableNames.map((i: YulNode) => this.transformYulIdentifier(i)),
+      );
+    }
+    const rhs = this.toSolidityExpr(node.value, lhs.typeString);
```

Comment:
Seems a little hackish to me. Is there a better solution? What about expressions that are not a part of an assignment